Update README and FAQ about containerizing mode (#1779)
GoogleContainerTools/jib: docs/faq.md

Normally, the plugin sets a default entrypoint for java applications, or lets you configure a custom entrypoint using the `container.entrypoint` configuration parameter. You can also override the default/configured entrypoint by defining a custom entrypoint when running the container. See [`docker run --entrypoint` reference](https://docs.docker.com/engine/reference/run/#entrypoint-default-command-to-execute-at-runtime) for running the image with Docker and overriding the entrypoint command, or see [Define a Command and Arguments for a Container](https://kubernetes.io/docs/tasks/inject-data-application/define-command-argument-container/) for running the image in a [Kubernetes](https://kubernetes.io/) Pod and overriding the entrypoint command.

### <a name="i-want-to-containerize-an-executable-jar"></a>I want to containerize a JAR.

The intention of Jib is to add individual class files, resources, and dependency JARs into the container instead of putting a JAR. This lets Jib choose an opinionated, optimal layout for the application on the container image, which also allows it to skip the extra JAR-packaging step.

> However, you can set `<containerizingMode>packaged` (Maven) or `containerizingMode = 'packaged'` (Gradle) to containerize a JAR, but note that your application will always be run via `java -cp ... your.MainClass` (even if it is an executable JAR). Some disadvantages:

Reviewer: `jib.containerizingMode = 'packaged'` for Gradle?

Author: Nice. Done.

### I want to containerize an executable JAR.

Currently Jib does not natively support creating an image that runs a runnable JAR (or WAR) through `java -jar runnable.jar` (although it is not impossible to configure Jib to do so at the expense of more complex project setup.) The intention of Jib is to add individual class files, resources, and dependency JARs into the container instead of putting a runnable JAR. This lets Jib choose an opinionated, optimal layout for the application on the container image, which also allows it to skip the extra JAR-packaging step.

> However, you can set `<containerizingMode>packaged` (Maven) or `containerizingMode = 'packaged'` (Gradle) to containerize a JAR, but note that your application will always be run via `java -cp your.MainClass` (even if it is an executable JAR). Some disadvnatags:

Reviewer 2: Should be `disadvantages` I guess.

Author: Oops, thanks!

- You need to run the JAR-packaging step (`mvn package` in Maven or the `jar` task in Gradle).

> - Reduced granularity in building and caching: if any of your Java source files or resource files are updated, not only the JAR has to be rebuilt but the layer containing the JAR in the image has to be recreated and pushed to a registry or a Docker daemon as a whole as a new layer.

Reviewer: Maybe something like:

```suggestion
- Reduced granularity in building and caching: if any of your Java source files or resource files are updated, not only the JAR has to be rebuilt, but the entire layer containing the JAR in the image has to be recreated and pushed to the destination.
```

- If it is a fat or shaded JAR embedding all dependency JARs, you are duplicating the dependency JARs in the image. Worse, it results in far more reduced granularity in building and caching, as dependency JARs can be huge and all of them need to be pushed repeatedly even if they do not change.

> ### I want to containerize a JAR.

Reviewer: Is there a reason for adding this? It seems redundant.

Author: Is there a way to change the title without changing the current link (`https://github.com/GoogleContainerTools/jib/blob/master/docs/faq.md#i-want-to-containerize-an-executable-jar`)? I wanted to make the old link continue to work.

Reviewer: I think you could try using just a normal anchor? We do something like that in the readmes.

`<a name="i-want-to-containerize-an-executable-jar"></a>### I want to containerize a JAR.`

Author: This works. Since I can change the title, I think I'll explain `<containerizingMode>` first and then explain runnable JARs later.

Author: Updated wordings and shuffled sentences. PTAL.
